modules/assim.sequential/R/Analysis_sda.R

```diff
@@ -0,0 +1,393 @@
+##' @export
+EnKF<-function(setting,Forcast,Observed,...){
+  
+  #------------------------------Setup
+  #-- reading the dots and exposing them to the inside of the function
+  dots<-list(...)
+  if (length(dots)>0) lapply(names(dots),function(name){assign(name,dots[[name]])})
+  for(i in seq_along(dots)) assign(names(dots)[i],dots[[names(dots)[i]]])
+
+    #Forcast inputs 
+  Q <- Forcast$Q # process error
+  Pf <- Forcast$Pf # forcast precision
+  mu.f <- Forcast$mu.f #mean forcast
+  X <- Forcast$X # states 
+  #Observed inputs
+  R <- Observed$R
+  Y <- Observed$Y
+  # Enkf---------------------------------------------------
+  ## design matrix
+  H <- matrix(0, length(Y), ncol(X)) #H maps true state to observed data
+  #linear
+  for (i in choose) {
+    H[i, i] <- 1
+  }
+  #non-linear fcomp
+  # for (i in choose) {
+  #   H[i, i] <- 1/sum(mu.f) #? this seems to get us on the right track. mu.f[i]/sum(mu.f) doesn't work. 
+  # }
+  ## process error
+  if (!is.null(Q)) {
+    Pf <- Pf + Q
+  }
+  
+  ## Kalman Gain
+  K <- Pf %*% t(H) %*% solve((R + H %*% Pf %*% t(H)))
+  # Analysis
+  mu.a <- mu.f + K %*% (Y - H %*% mu.f)
+  Pa   <- (diag(ncol(X)) - K %*% H) %*% Pf
+  return(list(mu.f = mu.f, Pf = Pf, mu.a = mu.a, Pa = Pa))
+}
+
+##' @title GEF
+##' @name  GEF
+##' @author Michael Dietze \email{dietze@@bu.edu}, Ann Raiho and Hamze Dokoohaki
+##' 
+##' @param settings  pecan standard settings list.  
+##' @param Forecast A list containing the forecasts variables including Pf (cov of forecast state variables), mu.f (vector of estimated mean of forecast state variables),
+##' Q (process variance) and X (a dataframe of forcats state variables for different ensemble)
+##' @param Observed A list containing the observed variables including R (cov of observed state variables) and Y (vector of estimated mean of observed state variables)
+##' @param ... Extra argument sent to the analysis function.
+##’ @details
+##’  
+##' 
+##' @description Given the Forecast and Observed this function performs the Generalized Ensemble Kalamn Filter. The generalized ensemble filter follows generally the three steps of sequential state data assimilation. But, in the generalized ensemble filter we add a latent state vector that accounts for added process variance. Furthermore, instead of solving the analysis analytically like the EnKF, we have to estimate the mean analysis vector and covariance matrix with MCMC.
+##' 
+##' @return It returns a list with estimated mean and cov matrix of forecast state variables as well as mean and cov estimated as a result of assimilation/analysis .
+##' @export
+GEF<-function(setting,Forcast,Observed,...){
+  #------------------------------Setup
+  #-- reading the dots and exposing them to the inside of the function
+  dots<-list(...)
+  if (length(dots)>0) lapply(names(dots),function(name){assign(name,dots[[name]], pos=1 )})
+  #General
+  var.names <- unlist(sapply(settings$state.data.assimilation$state.variable, 
```

Comment:
this could be reduced to `sapply(settings$state.data.assimilation$state.variable, '[[', "variable.name")`